File: Problems/Dynamic-programming/Combination-Sum-IV/Combination-Sum-IV.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

// Using top-down approach - Time: O(n * amount)

class Solution {
    std::unordered_map<int, int> m{{0, 1}};

    int dp(std::vector<int> &nums, int target)
    {
        if (m.count(target))
            return m[target];

        int cnt = 0;

        for (int n : nums)
        {
            if (n > target)
                break;
            cnt += dp(nums, target - n);
        }
        
        return m[target] = cnt;
    }

public:
    int combinationSum4(std::vector<int> &nums, int target)
    {
        std::sort(nums.begin(), nums.end());
        return dp(nums, target);
    }
};

int main()
{
    std::vector<int> nums = {1, 2, 3};
    int target = 4;

    std::cout << "Input: nums = [";
    for (int i = 0; i < nums.size(); i++)
    {
        std::cout << nums[i];
        if (i != nums.size() - 1)
            std::cout << ", ";
    }
    std::cout << "], target = " << target << std::endl;

    Solution sol;
    int result = sol.combinationSum4(nums, target);

    std::cout << "Output: " << result << std::endl;

    return 0;
}

File: Problems/Dynamic-programming/Combination-Sum-IV/Combination-Sum-IV.cs
using System;
using System.Collections.Generic;

// Using top-down approach - Time: O(n * amount)

public class Solution
{
    private Dictionary<int, int> m = new Dictionary<int, int> { { 0, 1 } };

    private int Dp(int[] nums, int target)
    {
        if (m.ContainsKey(target)) return m[target];

        int cnt = 0;

        foreach (int n in nums)
        {
            if (n > target) break;
            cnt += Dp(nums, target - n);
        }
        m[target] = cnt;

        return cnt;
    }

    public int CombinationSum4(int[] nums, int target)
    {
        Array.Sort(nums);
        return Dp(nums, target);
    }
}
class Program
{
    static void Main(string[] args)
    {
        int[] nums = { 1, 2, 3 };
        int target = 4;

        Console.Write("Input: nums = [");
        for (int i = 0; i < nums.Length; i++)
        {
            Console.Write(nums[i]);
            if (i < nums.Length - 1)
            {
                Console.Write(", ");
            }
        }
        Console.WriteLine("], target = " + target);

        Solution sol = new Solution();
        int result = sol.CombinationSum4(nums, target);

        Console.WriteLine("Output: " + result);
    }
}
